linkup-cli/src/start.rs

```diff
@@ -1,165 +1,231 @@
-        if sys.file_exists(Path::new(&file_path)) {
-            log::info!("Tunnel config file for {}: {}", tunnel_id, file_path);
-        } else {
-            log::info!("Tunnel config file for {} does not exist", tunnel_id);
-            create_tunnel = true;
-        }
-    }
+        LocalState::load()?
+    };
 
-    if create_tunnel {
-        println!("Creating tunnel...");
-        tunnel_id = paid_manager.create_tunnel(&tunnel_name)?;
-        paid_manager.create_dns_record(&tunnel_id, &tunnel_name)?;
+    let status_update_channel = sync::mpsc::channel::<services::RunUpdate>();
+
+    let local_server = services::LocalServer::new();
+    let cloudflare_tunnel = services::CloudflareTunnel::new(state.linkup.session_name.clone());
+    let caddy = services::Caddy::new(state.domain_strings());
+    let dnsmasq = services::Dnsmasq::new(state.linkup.session_name.clone(), state.domain_strings());
+
+    let mut display_thread: Option<JoinHandle<()>> = None;
+    let display_channel = sync::mpsc::channel::<bool>();
+
+    // If we are doing RUST_LOG=debug to debug if there is anything wrong, having the display thread make so it
+    // overwrites some of the output since it does some cursor moving.
+    // So in that case, we do not start the display thread.
+    if !log::log_enabled!(log::Level::Debug) {
+        display_thread = Some(spawn_display_thread(
+            &[
+                services::LocalServer::NAME,
+                services::CloudflareTunnel::NAME,
+                services::Caddy::NAME,
+                services::Dnsmasq::NAME,
+            ],
+            status_update_channel.1,
+            display_channel.1,
+        ));
     }
 
-    if tunnel_manager.is_tunnel_running().is_err() {
-        println!("Starting paid tunnel...");
-        state.linkup.tunnel = Some(tunnel_manager.run_tunnel(&state)?);
-    } else {
-        println!("Cloudflare tunnel was already running.. Try stopping linkup first if you have problems.");
+    // To make sure that we get the last update to the display thread before the error is bubbled up,
+    // we store any error that might happen on one of the steps and only return it after we have
+    // send the message to the display thread to stop and we join it.
+    let mut exit_error: Option<Box<dyn std::error::Error>> = None;
+
+    match local_server
+        .run_with_progress(&mut state, status_update_channel.0.clone())
+        .await
+    {
+        Ok(_) => (),
+        Err(err) => exit_error = Some(Box::new(err)),
     }
-    state.save()?;
 
-    if sys.file_exists(&linkup_file_path(LINKUP_LOCALDNS_INSTALL)) {
-        boot.boot_local_dns(state.domain_strings(), state.linkup.session_name.clone())?;
+    if exit_error.is_none() {
+        match cloudflare_tunnel
+            .run_with_progress(&mut state, status_update_channel.0.clone())
+            .await
+        {
+            Ok(_) => (),
+            Err(err) => exit_error = Some(Box::new(err)),
+        }
```

Comment:
Out of curiosity: they all run with progress using the same arguments or? What was the reason you moved away from our initial loop implementation?

Parallelising these now that we have tokio in a next step would be nice, we would have to provide a lock between tunnel and dns setup to enforce ordering, but feels doable.

Reply:
> they all run with progress using the same arguments or? What was the reason you moved away from our initial loop implementation?

Yeah, this is a slight type issue. The problem here is size calculation at compile time.
Since the array would need to have some kind of (invalid) `Vec<dyn BackgroundService<dyn std::error::Error>>` because each service have a different error type. And that doesn't seem to be possible.
I think it could work if all background services had the same error struct, but I think is nice that they are independent and correctly typed 🤔 
Wdyt?

Reply:
Or maybe if we box it all. But then I think the implementation gets... weird. Not sure if we would gain that much.
But will take a look into it!

Reply:
Let's leave it like this for now - either we will run into it when we try to run things in parallel or we won't - I agree that the error ergonomics are nice. Here's a little rust gets hairy example eh? 😅